File: day2/sonar_direction.py
def part1(moves):
    h_pos = 0
    depth = 0

    for direction, unit in moves:
        if direction == 'forward':
            h_pos += int(unit)
        elif direction == 'up':
            depth -= int(unit)
        elif direction == 'down':
            depth += int(unit)

    return h_pos * depth


def part2(moves):
    aim = 0
    h_pos = 0
    depth = 0

    for direction, unit in moves:
        if direction == 'forward':
            h_pos += unit
            depth += aim * unit
        elif direction == 'up':
            aim -= unit
        elif direction == 'down':
            aim += unit

    return h_pos * depth


def main():
    # Read in the data.
    with open('./day2/input.txt', 'r') as f:
        moves = f.read().split('\n')

    # Ignore this final newline in the input. Frustrating.
    if not moves[-1]:
        moves = moves[:-1]

    # Parse the data in the form (direction, unit).
    parsed_moves = [
        (move.split(' ')[0], int(move.split(' ')[1]))
        for move in moves
    ]

    print(part1(parsed_moves))
    print(part2(parsed_moves))


if __name__ == '__main__':
    main()

File: day2/sonar_direction.rb
inputData = File.read("input.txt").split

# Part 1
$horizontal = 0
$depth = 0

inputData.each_slice(2) {|a, b| a == "forward" ? $horizontal += b.to_i : a == "up" ? $depth -= b.to_i : $depth += b.to_i}

puts $horizontal * $depth

# Part 2
$horizontal = 0
$depth = 0
$aim = 0

def add(num)
    $horizontal += num
    $depth += $aim * num
end

inputData.each_slice(2) {|a, b| a == "forward" ? add(b.to_i) : a == "up" ? $aim -= b.to_i : $aim += b.to_i}

puts $horizontal * $depth